### PR TITLE
starting to add original leaf tests over

### PR DESCRIPTION
**Pull request by loganwright**

### CODE REVIEW

File: Sources/LeafKit/LeafSyntax.swift [hunk 1]
```diff
@@ -79,19 +79,19 @@ indirect enum LeafSyntax: CustomStringConvertible {
         switch self {
         case .raw(var byteBuffer):
             let string = byteBuffer.readString(length: byteBuffer.readableBytes) ?? ""
-            return string.debugDescription
+            return "raw: " + string.debugDescription
```
Comment (loganwright) [inline Sources/LeafKit/LeafSyntax.swift]: this is just temporary for me, I was trying to better visualize my logs

File: Sources/LeafKit/Exports.swift [hunk 2]
```diff
@@ -5,6 +5,7 @@ extension UInt8 {
     static let quote: UInt8 = 0x22
     static let octothorpe: UInt8 = 0x23
     static let leftParenthesis: UInt8 = 0x28
```
Comment (tanner0101) [inline Sources/LeafKit/Exports.swift]: We should convert the rest of these to the `.utf8.first!` syntax. So much better.

File: Sources/LeafKit/LeafLexer.swift [hunk 3]
```diff
@@ -1,147 +1,210 @@
-struct LeafLexer {
-    private var buffer: ByteBuffer
+extension ByteBuffer {
+    mutating func stringify() -> String? {
```
Comment (tanner0101) [inline Sources/LeafKit/LeafLexer.swift]: Prefixing this with `read...` is NIO's pattern for letting you know it's a mutating method (increments the reader index). Maybe `readRestAsString()` or `readEntireString()`?

Non-mutating methods are `get...` (and you pass in an explicit reader index).

File: Sources/LeafKit/LeafParser.swift [hunk 4]
```diff
@@ -19,13 +19,20 @@ struct LeafParser {
...
         switch peek {
         case .raw(let raw):
             self.pop()
             return .raw(raw)
         case .tag(let name):
             self.pop()
             return self.nextTag(named: name)
+        case .tagIndicator:
+            self.pop()
+            if self.peek() == LeafToken.parametersStart {
```
Comment (loganwright) [inline Sources/LeafKit/LeafParser.swift]: vestigial, rm

File: Sources/LeafKit/LeafParser.swift [hunk 5]
```diff
@@ -19,13 +19,20 @@ struct LeafParser {
         guard let peek = self.peek() else {
             return nil
         }
+        print("peek: \(peek)")
```
Comment (loganwright) [inline Sources/LeafKit/LeafParser.swift]: rm

File: Sources/LeafKit/LeafParser.swift [hunk 6]
```diff
@@ -180,6 +187,9 @@ struct LeafParser {
                 self.pop()
                 return self.nextTag(named: name)
             }
+        case .tagIndicator:
```
Comment (loganwright) [inline Sources/LeafKit/LeafParser.swift]: rm

File: Sources/LeafKit/LeafLexer.swift [hunk 7]
```diff
@@ -1,147 +1,210 @@
...
+        return readString(length: readableBytes)
+    }
+}
+
+extension UInt8 {
+    var isValidInTagName: Bool {
+        return isLowercaseLetter
```
Comment (tanner0101) [inline Sources/LeafKit/LeafLexer.swift]: For Vapor 4, I've been adopting some things from NIO's code style like always using `self.`. It makes the code a lot easier to reason about in places without context sensitive syntax-highlighting (like GitHub)

File: Sources/LeafKit/LeafLexer.swift [hunk 8]
```diff
@@ -1,147 +1,210 @@
...
+                // consume '\' only in event of '\#'
+                // otherwise allow it to remain for other
+                // escapes, a la javascript
+                if peek(at: 1) == .octothorpe {
+                    pop()
                 }
-                self.state = .tag
-                return .tag(name: name)
+                // either way, add raw '#' or '\' to registry
+                return buffer.readSlice(length: 1).map(LeafToken.raw)
```
Comment (tanner0101) [inline Sources/LeafKit/LeafLexer.swift]: i think after consuming the `\` you'd want to return `self.nextToken()` recursively to parse out the tag

Comment (tanner0101) [reply Sources/LeafKit/LeafLexer.swift]:
```swift
if self.peek(at: 1) == .octothorpe {
    self.pop() // remove \
    return self.nextToken() // now parse tag
} else {
    // this is just a normal \
    return buffer.readSlice(length: 1).map(LeafToken.raw) 
}
```

File: Sources/LeafKit/LeafLexer.swift [hunk 9]
```diff
@@ -1,147 +1,210 @@
...
-                self.state = .parameters
-                return .parametersStart
-            case .colon:
-                self.buffer.moveReaderIndex(forwardBy: 1)
-                self.state = .normal
-                return .tagBodyIndicator
+                // '#('
+                state = .parameters(depth: 0)
+                return .tag(name: "")
+            case let x where x.isValidInTagName:
```
Comment (tanner0101) [inline Sources/LeafKit/LeafLexer.swift]: this is nice 👍

File: Sources/LeafKit/LeafLexer.swift [hunk 10]
```diff
@@ -1,147 +1,210 @@
...
-                guard let name = self.buffer.readString(length: length) else {
-                    return nil
-                }
+                pop()
+                return try self.nextToken()
+            case let x where x.isValidInParameter:
+                let read = readWhile { $0.isValidInParameter }
+                guard let name = read else { fatalError("switch case should disallow this") }
+                // this parameter is a tag
+                if peek() == .leftParenthesis { return .tag(name: name) }
```
Comment (tanner0101) [inline Sources/LeafKit/LeafLexer.swift]: guard might be nice here to ensure return

File: Sources/LeafKit/LeafLexer.swift [hunk 11]
```diff
@@ -1,147 +1,210 @@
...
-    func peek() -> UInt8? {
-        return self.buffer.getInteger(at: self.buffer.readerIndex)
+    mutating func readSliceWhile(_ check: (UInt8) -> Bool) -> ByteBuffer? {
+        guard let length = countMatching(check: check) else {
+            return nil
+        }
+        return buffer.readSlice(length: length)
+    }
+    
+    func peek(at idx: Int = 0) -> UInt8? {
```
Comment (tanner0101) [inline Sources/LeafKit/LeafLexer.swift]:
```suggestion
    func peek(aheadBy offset: Int = 0) -> UInt8? {
```

File: Sources/LeafKit/LeafLexer.swift (hunk 8 above)
Comment (loganwright) [reply Sources/LeafKit/LeafLexer.swift]: we need to stack the `#` key away or the lexer will want to make it a tag in this case 👍

File: Sources/LeafKit/LeafLexer.swift (hunk 11 above)
Comment (loganwright) [reply Sources/LeafKit/LeafLexer.swift]: was stuck what to name this, aheadBy is way better

**Comment by penny-coin** [comment]
Hey @loganwright, you just merged a pull request, have a coin! 

You now have 518 coins.

**Pull request merged**